### DISCUSSION TITLE
User can have many pages

### DISCUSSION THREAD

**Discussion by rudzainy** [rudzainy/ranting#97] [Ideas]
### Structure
 - Header image
 - Title (user's default profile page will use user's full name)
 - Default image (user's default profile page will use user's avatar)
 - Description (user's default profile page will use user's description)
 - Image gallery
 - Groups of Links and Pages
 
Alternatively, migrate user's full name, avatar and description into Page, and remove columns in user model)

### Current user journey
Create account (devise) > create profile page (user model)

### New user journey
Create account (devise) > create profile page (page model)

## Thoughts dump

```
class User < ApplicationRecord
  ...
  has_one_attached :avatar
  has_one_attached :qr_code
  has_one :page, dependent: :destroy # assign default profile page
  ...
end
```
Removed `has_many :links, dependent: :destroy`. Links from current users will be migrated under user's default profile page.

```
class Page < ApplicationRecord
  ...
  has_many :groups, dependent: :destroy
  belongs_to :user
  ...
end
```

```
class Group < ApplicationRecord
  # TODO: Groups should be sharable
  belongs_to :created_by, class_name: "Page", foreign_key: "page_id"
  has_many :links
end
```

```
class Link < ApplicationRecord
  # TODO: Break down features in links

  is_impressionable
  has_one_attached :image
  has_one_attached :qr_code
  belongs_to :page, optional: true
  belongs_to :group, optional: true
  ...
end
```

**Comment by rudzainy** [comment]
https://spinacms.com/docs/getting-started/existing-project

https://spinacms.com/code-examples

**Reply by rudzainy** [reply]
@hewrin ☝️

**Reply by rudzainy** [reply]
we can just use a cms for page model. I'm thinking of using the cms' UI for managing QR and short links also. sound too bootstrapy tho...

**Comment by hewrin** [comment]
What are these pages? Mcm blog ke?

**Reply by rudzainy** [reply]
think of it as separating profile page in user model into it's own model. 

if we generalise user profile page, it's just a fixed-structure static page, where users can add groups of links. I'm thinking of extending that capability to users. they can use it as a blog, subpages, product page, marketing page etc.

one control that we should consider is to set a limit to how deep we allow users to link to a subpage from user's profile page. I say one level deep is enough, two levels at most 😬

**Reply by rudzainy** [reply]
| Figma file | https://www.figma.com/design/tT3w4fhvBSz7PYXfRhuLHF/Gugel.my?node-id=80-243097&t=aovWtrnCNjRqPzoz-1 |
|---|---|
| Figma prototype | https://www.figma.com/proto/tT3w4fhvBSz7PYXfRhuLHF/Gugel.my?page-id=80%3A243097&node-id=2028-29193&viewport=372%2C677%2C0.18&t=2fmVgGDQrQjBKna1-1&scaling=min-zoom&content-scaling=fixed&starting-point-node-id=2028%3A29193&show-proto-sidebar=1 |
<img width="1288" alt="Screenshot 2024-07-28 at 1 14 29 PM" src="https://github.com/user-attachments/assets/d2cfc9c4-eee2-4932-a127-fe80db8bbc63">

<img width="1385" alt="Screenshot 2024-08-17 at 12 00 21 PM" src="https://github.com/user-attachments/assets/c827322b-8183-41e9-b49e-75a73e149c4e">
